### DISCUSSION TITLE
eth_getProof on zk-sync

### DISCUSSION THREAD

**Discussion by AlexNi245** [zkSync-Community-Hub/zksync-developers#273] [General] [answered]
### Environment

Testnet

### Provide a brief description of the functionality you're trying to implement.

Ethereum mainnet offers the "eth_getProof" RPC call that can be used to compute a storage proof for arbitrary data on-chain. For Rollups like Optimism or Arbitrum this storage proof can be used to proof that certain data is stored on that layer2 by only having access to Ethereum.

I'm looking for a similar feature on zy-sync but according to the docs "eth_getProof" is not supported yet by zy-sync

### What is the specific issue or error you're encountering?

"eth_getProof" is not availbale

### Can you share the error messages or logs you're receiving, if any?

-

### Have you made any recent changes to the contract before encountering this issue?

-

### Are there any external libraries or contracts that your contract interacts with?

-

### Can you provide the relevant portions of your contract code where the issue is occurring?

-

### Have you tried to isolate the problem, and if so, what were the results?

-

### What steps have you already taken to try to resolve the issue?

*Studied the API docs
*Studied SDK methods
Found rpc methods zks_getL2ToL1LogProof and zks_getL2ToL1LogProof but not sure if they can be used to compute storage proofs.

**Comment by uF4No** [comment]
Hi @AlexNi245 

We don't have `eth_getProof` but you can use `zks_getProof` instead. I just realised that this method is not documented in our API docs yet so we'll add it as soon as possible.

**Reply by AlexNi245** [reply]
@uF4No can you link me to any resource that explains how to verify the result of zks_getProof onchain ?

**Reply by mm-zk** [reply, marked as answer]
@AlexNi245 - AFAIK we don't have the merkle proofs verification implemented in solidity.

What I can show you,  is my hobby project - https://github.com/mm-zk/zksync_tools/blob/main/tx_prover/merkle_proof.py#L30 - where I verify it in python.

**Reply by AlexNi245** [reply]
@mm-zk thanks for sharing your project, that was exactly what I was looking for. Will try to rewrite it for TypeScript and Solidity soon.

**Comment by AlexNi245** [comment]
Awesome, I tested it out and it works with the same arguments as eth_getProof. However, the response varies since it doesn't contain a storage hash of the Smart-Contract account. I guess that is because the result is not a Merkle-proof but something different.

Also, how could I verify that result? I've seen there is a contract called Verifier on L1. Is that the right way to verify the zks proof?

**Reply by uF4No** [reply]
We just added the `zks_getProof` info to [the API documentation](https://docs.zksync.io/build/api.html#zks-getproof), including the zkSync Merkle tree specification.

For verification, please check the message from @mm-zk as a reference.

**Reply by AlexNi245** [reply]
Thanks for providing the exhaustive explanation in the API docs !

**Comment by belanoparo** [comment]
Currently, zkSync does not support the "eth_getProof" RPC call like Ethereum mainnet does. However, zkSync provides alternative methods such as "zks_getL2ToL1LogProof" and "zks_getL2ToL1LogProof" which can be used to obtain proofs related to transactions between L2 and L1. These methods may not directly provide storage proofs, but they can be used to verify transactions and data integrity on zkSync. For specific storage proofs, you may need to explore other approaches or wait for zkSync to potentially add support for such functionality in the future.

**Comment by sarportzapbisusan** [comment]
Currently, zksync does not support the "eth_getProof" RPC call like Ethereum mainnet does. However, there are alternative methods such as "zks_getL2ToL1LogProof" and "zks_getL2ToL1LogProof" available, but they may not directly compute storage proofs. Further investigation and potentially reaching out to the zksync team for clarification or alternative solutions may be necessary.

**Comment by lykinsjennifer** [comment]
Currently, zksync does not support the "eth_getProof" RPC call for computing storage proofs similar to Ethereum mainnet. Alternative methods such as "zks_getL2ToL1LogProof" and "zks_getL2ToL1LogProof" are available but they are specifically designed for obtaining proofs related to L2-to-L1 transactions and logs. Further development or updates from the zksync team may introduce support for storage proofs in the future.

**Comment by colleennealv** [comment]
You can interact with smart contracts on zkSync using the Python SDK by deploying contracts, calling their functions, and querying their states. You may refer to the zkSync Python SDK documentation and explore examples provided there for guidance.

**Comment by knowenlophoroger** [comment]
The user is trying to implement a functionality that requires computing a storage proof for arbitrary data on-chain using zksync, but the "eth_getProof" RPC call is not supported by zksync yet. The user has studied the API docs and SDK methods, but is not sure if the zks_getL2ToL1LogProof and zks_getL2ToL1LogProof RPC methods can be used to compute storage proofs. No error messages, logs, or relevant portions of contract code are provided, and no repository link or additional details are given.
